### PR TITLE
fix(ci): reduce mixtral release smoke batch

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Reduce the Mixtral 8x7B SQuAD release CI smoke workload so the job no longer OOMs during the optimizer gradient-clipping all-reduce, and use THD packed-sequence collation for the recipe to keep pipeline sequence shapes fixed at the packed size.

# Changelog

- Increase the Mixtral SQuAD CI timeout from `00:20:00` to `00:30:00` to account for model setup plus the smoke run.
- Move the memory-reduction settings into their recipe sections: `step_scheduler.global_batch_size=64`, `step_scheduler.local_batch_size=8`, and `distributed.pipeline.pp_microbatch_size=1`.
- Keep `ci.release.step_scheduler.max_steps=5` because release defaults otherwise override recipe `max_steps` to 50.
- Enable THD packed-sequence collation for train and validation with `packed_sequence_size=1024`.
- Pack validation when the validation dataloader uses the THD collater and packed sequences are enabled, while preserving main's backend/model opt-in path for THD validation packing.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests? Added focused validation-dataloader tests for THD validation packing.
- [x] Did you add or update any necessary documentation? Not needed for CI recipe/runtime behavior.

# Additional Information

Latest PR head after rebasing on `origin/main`: `8b0a5c6d`.

Local validation on commit `8b0a5c6d`:

- `uv run pytest -q tests/unit_tests/recipes/test_train_ft.py -k 'build_validation_dataloader'`
- `uv run ruff check nemo_automodel/recipes/llm/train_ft.py tests/unit_tests/recipes/test_train_ft.py`
- `uv run ruff format --check nemo_automodel/recipes/llm/train_ft.py tests/unit_tests/recipes/test_train_ft.py`

nemo-ci validation before the rebase, on equivalent pre-rebase commit `90063926`:

- nemo-ci root pipeline: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55521454
- AutoModel child pipeline: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55522027
- LLM child pipeline: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55522191
- Target Mixtral job: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/345692983
- Result: success. Slurm completed with `SLURM_STATE=COMPLETED` and `SLURM_EXITCODE=0`.
- Trace check: only two `Precomputed pipeline stage shapes` messages, both `seq_len=1024`; no variable validation sequence-shape spam.

Earlier validation:

- Commit `2df24c4d` nemo-ci target job passed but exposed validation still using variable sequence lengths: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/345669255
- Commit `a143805d` initial memory-reduction validation: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/345494195

### CODE REVIEW

**Comment by akoumpa** [comment]
Latest nemo-ci rerun for commit `90063926` passed.

- Root pipeline: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55521454
- AutoModel child: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55522027
- LLM child: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/55522191
- Target Mixtral job: https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/jobs/345692983

Trace check: Slurm completed with `SLURM_STATE=COMPLETED` and `SLURM_EXITCODE=0`. There were only two `Precomputed pipeline stage shapes` messages and both were `seq_len=1024`, so the variable validation shape spam is gone.

**Comment by akoumpa** [comment]
Reasoning for removing the Magi/TE backend allowlist from validation packing:

The validation decision here is whether to pass `cfg_ps` into `build_dataloader()` so the validation dataset gets packed. That should be based on the dataloader contract, not on a backend allowlist:

- packed sequences are enabled: `packed_sequence.packed_sequence_size > 0`
- validation uses `packed_sequence_thd_collater`, which consumes THD/cu_seqlens metadata

The previous Magi/TE check was too narrow for this case. Mixtral was configured with THD packed collation, so training packed correctly, but validation skipped packing because the backend allowlist evaluated false. That is why the nemo-ci trace still had variable validation shape logs like `seq_len=310` after the config-only change.

This does not remove the actual safety check. `build_dataloader()` still checks whether the model can consume `seq_lens`; if not, it logs a warning and disables packed sequence. So Magi remains supported, but it is no longer the gate for validation packing. The gate is now: packed sequence requested + THD validation collater present, with model support still validated inside `build_dataloader()`.

The rerun on `90063926` confirmed the behavior: only two `Precomputed pipeline stage shapes` messages, both `seq_len=1024`, and the nemo-ci Mixtral job passed.

**Comment by akoumpa** [comment]
Rebased on current `origin/main` and resolved the `train_ft.py` conflict from `0c342ab4`.

Resolution kept both paths:

- explicit validation THD collater + `packed_sequence_size > 0`, which fixes the Mixtral validation shape spam
- the main-branch backend/model opt-in path: `magi`, TE, or `should_pack_validation_with_training()` when training uses THD

GitHub now reports the PR as mergeable: `mergeable=true`; current head is `8b0a5c6d`. Local focused validation passed:

- `uv run pytest -q tests/unit_tests/recipes/test_train_ft.py -k "build_validation_dataloader"`
- `uv run ruff check nemo_automodel/recipes/llm/train_ft.py tests/unit_tests/recipes/test_train_ft.py`
- `uv run ruff format --check nemo_automodel/recipes/llm/train_ft.py tests/unit_tests/recipes/test_train_ft.py`
